### PR TITLE
Storing metadata information in GetResult

### PR DESCRIPTION
**Pull request by sandmannn**
This PR adds a new parameter to the constructor of document field `isMetadataField` which describes whether the DocumentField in question is metadata field or not. It is part of larger refactoring that aims to remove the calls to static methods of `MapperService` related to metadata fields, as discussed in elastic#24422.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 1]
```diff
@@ -350,15 +352,15 @@ public static GetResult fromXContentEmbedded(XContentParser parser, String index
                     }
                 } else if (FIELDS.equals(currentFieldName)) {
                     while(parser.nextToken() != XContentParser.Token.END_OBJECT) {
-                        DocumentField getField = DocumentField.fromXContent(parser);
+                        DocumentField getField = DocumentField.fromXContent(parser, false);
```
Comment (Sandmannn) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: @rjernst This DocumentField is not inside of `_source`, but I assumed that it is not a metadata field based on the way the FIELDS component is filled during serialization https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/get/GetResult.java#L277-L282

**Comment by mayya-sharipova** [comment]
@Sandmannn Thank you for your PR.  Can you please sing CLA, so after that we can start CI test for this PR.

**Comment by Sandmannn** [comment]
Hi, I actually signed it several times, but the check is still red. Should I provide a Transaction ID?

**Comment by javanna** [comment]
sorry @Sandmannn it looks like we had some technical problems at the time that you signed the CLA, we apologize. Would you mind signing it once again? Hopefully this time it will work.

**Comment by Sandmannn** [comment]
Hi @javanna , I just did it again. looks like that check is passed now.

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by elasticcla** [comment]
Hi @Sandmannn, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch/pull/38373.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

File: server/src/main/java/org/elasticsearch/common/document/DocumentField.java [hunk 2]
```diff
@@ -47,14 +46,16 @@
 public class DocumentField implements Streamable, ToXContentFragment, Iterable<Object> {
 
     private String name;
+    private Boolean isMetadataField;
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/common/document/DocumentField.java]: can this be `final`?

**Review by mayya-sharipova — CHANGES_REQUESTED** [review]
@Sandmannn thanks for your PR, here is the 1st round of requested changes.

File: server/src/main/java/org/elasticsearch/common/document/DocumentField.java [hunk 3]
```diff
@@ -132,7 +133,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
-    public static DocumentField fromXContent(XContentParser parser) throws IOException {
+    public static DocumentField fromXContent(XContentParser parser, boolean inMetadataArea) throws IOException {
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/common/document/DocumentField.java]: `inMetadataArea` why not call this parameter the same `isMetadataField`?

File: server/src/main/java/org/elasticsearch/common/document/DocumentField.java (hunk 2 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/common/document/DocumentField.java]: We need to think more how to organize `DocumentField` class.  Usually `fromXContent -> toXContent ->fromXContent` should produce an equal object (and this is what we test in `DocumentFieldTests`). The way how you organized a `DocumentField` class, it doesn't happen as `toXContent` is using `isMetadataField`, and `fromXConent` is not using it.

Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/common/document/DocumentField.java]: if it requires a lot of changes to use `isMetadataField` in `toXContent` (we need to investigate this first), then we should document why we are not including it in `toXContent`, and also exclude  this field from `equals`, `hashCode` - also documenting why `isMetadataField`  is not participating in these functions

Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/common/document/DocumentField.java]: Thanks for the review! 

Regarding consistency of ```fromXContent -> toXContent ->fromXContent ``` we had a relevant discussion in the issue description. There was a suggestion of using different json structures  depending when serializing from new versions, while still being able to deserialize json from old versions, as suggested in https://github.com/elastic/elasticsearch/issues/24422#issuecomment-455913816  . In such case it would have been possible to have enough information in the json to be able to make the conversions consistent without context. Yet there was a strong argument to avoid as mentioned here https://github.com/elastic/elasticsearch/issues/24422#issuecomment-456632130
as changing such json structure may result in issues when serializing it in one version and parsing in another.

In short, looks like it is not a lot of changes, but a bwc concern that prevents us from storing the `isMetadataField` in `toXContent`. It would be great if you share your thoughts on the matter.

File: server/src/main/java/org/elasticsearch/common/document/DocumentField.java (hunk 3 above)
Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/common/document/DocumentField.java]: this parameter naming follows the intent to emphasize and make it explicit that we decide whether the field is metadata or not depending on whether it is located in metadata area of xcontent or not. 

I don't really have a strong opinion here, we can replace it with isMetadataField and add some comments, if you prefer that.

File: server/src/main/java/org/elasticsearch/common/document/DocumentField.java (hunk 2 above)
Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/common/document/DocumentField.java]: here we need to make a trade-off between a)extending serialized json schema for document field, additional coding to guarantee backward compatibility for serialized content and b)guessing the meaning of json fields depending on the context, which may potentially introduce some bugs if our assumptions are incorrect in some case, e.g. it is not completely obvious if it is a right approach here https://github.com/elastic/elasticsearch/pull/38373/files/314c58a64e5c4f2fb3fc9ec6e3366a0209597dd5#r253675646 

what is the process for deciding between `a` and `b` here?

File: server/src/main/java/org/elasticsearch/common/document/DocumentField.java (hunk 3 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/common/document/DocumentField.java]: thanks for the explanation, `isMetadataField` sounds better for consistency with the rest of the code. And making a field being a part of medata makes it a metadata field.

File: server/src/main/java/org/elasticsearch/common/document/DocumentField.java (hunk 2 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/common/document/DocumentField.java]: @Sandmannn Thanks for clarification. Ok, lets leave your code for json serialization for `DocumentField` at it is now, but just put comments why `isMetadataField` is a special field, why it is not participating in equals, hashCode, toString, and `toXContent`.
You still need to modify `readFrom` and `writeTo` methods to include this field.

**Review by rjernst — COMMENTED** [review]
After looking through this in more detail, I wonder if DocumentField should not have any `isMetadataField()` method at all. Currently we determine whether a field is a meta field or not through this static list in the MapperService. Yet we want this to be through the registration of the underlying Mapper (which we already have, it is just not exposed to most parts of the code). DocumentField itself is a POJO of sorts, having no knowledge of the node it is being serialized/deserialized on.  Instead of trying to force this information into the DocumentField, I think we can modify the two callers of `DocumentField.isMetadataField()` to look elsewhere (ie the MapperService) to determine whether a field is a meta field or not. This would have the same underlying effect: the node the DocumentField is being used on would control where in the structure the field ends up. But this is no worse than the node level static list we have today, and matches with the node level registration we have.

**Comment by Sandmannn** [comment]
Thanks for the suggestion @rjernst , it might result in a cleaner solution. Yet currently it returns us to the old question from the issue discussion https://github.com/elastic/elasticsearch/issues/24422#issuecomment-446302244 i.e. what can be the right way of getting the right instance of `MapperService` in the xContent serialization functions of `GetResult` and `SearchHit` where we need this information. Is there any way to directly access the current `Node` and/or  `MapperService` ? Otherwise we may need to inject `MapperService` in all constructors of `GetResult` and `SearchHit`, which is also quite messy.

**Comment by rjernst** [comment]
@Sandmannn You are right my suggestion raises other concerns. I looked over this with @mayya-sharipova and we came up with a slightly different approach I think will be much cleaner. The idea is to change the places storing DocumentField to have two maps, one for fields and one for metadata fields. In all the DocumentField construction sites you changed here, the object is placed into a Map which later gets read when serializing xcontent in GetResult and SearchHit. By those two classes use two maps, the call sites you changed here can decide which map to place the newly constructed DocumentField into, rather than stashing this knowledge inside the DocumentField. I don't think it would be a very large change to this PR, but if it turns out to be, adding the new map could be a done as a first isolated PR.

**Comment by Sandmannn** [comment]
Thanks for aligning @rjernst , @mayya-sharipova ! Just to make sure I got the idea right, let me paraphrase it. Lets focus for a moment at `GetResult` as it already has a construct that looks somewhat similar. When we are serializing `GetResult` we are already separating the metadata fields from other fields (currently not in maps, but in lists) here https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/get/GetResult.java#L254-L257 

If I understanding the suggestion correctly, the idea is to do this separation even earlier, i.e. when we are receiving the fields in the constructor, we would separate them into something like `this.metadataFieldsMap` and `this.otherFieldsMap` instead of filling the `this.fields` directly
https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/get/GetResult.java#L93-L95

In such case, I think one piece is still not completely clear for me: how will we separate between metadata and non-metadata fields in constructor of GetResult? Using the same `field.isMetadataField()` (which I believe is not our intent) or via something else? Once we have this separating information I think we will be able to keep it during serialization/deserialization, but getting this separation in the first place may be a bit tricky.

**Comment by rjernst** [comment]
We wouldn't be doing the split within the `GetResult` constructor. It would be at the locations calling the constructor. In the case of `GetResult`, this is deserialization, and ShardGetService. For the first, we need to update the serialization format for GetResult to store two separate maps, and in deserialization we would use the existing static method (it will have to remain in the code until we no longer support rolling upgrades to whichever version this is added in, ie 8.0). For ShardGetService, we can have the MapperService.

**Comment by Sandmannn** [comment]
Changed the focus of this PR from DocumentField into GetResults. Would be nice to get some general confirmation that the content of this PR conforms to the suggested design before I do further adjustments in test.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 4]
```diff
@@ -68,7 +69,8 @@
     private long seqNo;
     private long primaryTerm;
     private boolean exists;
-    private Map<String, DocumentField> fields;
+    private Map<String, DocumentField> otherFields;
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: I think keeping "fields", or calling this "documentFields" would be good. "other" fields sounds like this are something extraneous, when in fact they are the main fields.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 5]
```diff
@@ -224,19 +230,24 @@ public String sourceAsString() {
     }
 
     public Map<String, DocumentField> getFields() {
+        Map<String, DocumentField> fields = emptyMap();
+        fields.putAll(metaFields);
+        fields.putAll(otherFields);
+        // TODO: question to reviewers - do you think it makes sense to cache this merged `fields` map?
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: I wouldn't worry about caching for now. Separately from this PR we should look at changing all the consumers of this method.

**Review by rjernst — COMMENTED** [review]
Sorry I lost track of this. At a high level this is looking better. I would appreciate @mayya-sharipova's comments as well.

**Comment by Sandmannn** [comment]
thanks a lot, I will tag you next time I create a new commit in this PR.

**Comment by Sandmannn** [comment]
@rjernst @mayya-sharipova  updated PR according to last comments, looking forward to further reviews.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 5 above)
Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: ok

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 4 above)
Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: renamed for `documentFields`

**Comment by rjernst** [comment]
@elasticmachine ok to test

File: server/src/test/java/org/elasticsearch/action/get/GetResponseTests.java [hunk 6]
```diff
@@ -64,9 +64,16 @@ private void doFromXContentTestWithRandomFields(boolean addRandomFields) throws
         Tuple<GetResult, GetResult> tuple = randomGetResult(xContentType);
         GetResponse getResponse = new GetResponse(tuple.v1());
         GetResponse expectedGetResponse = new GetResponse(tuple.v2());
+        if (getResponse.getResult.getMetadataFields().containsKey("_ignored")) {
+//            System.out.println("mew");
+            int a = 3;
+            int b = a-4;
+        }
```
Comment (Sandmannn) [inline server/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: sorry, missed to clean some debug coding, will remove it in next update.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 7]
```diff
@@ -222,20 +226,34 @@ public String sourceAsString() {
...
 
+
+    public Map<String, DocumentField> getMetadataFields() {
+        return new HashMap<>(metaFields);
+    }
+
+    public Map<String, DocumentField> getDocumentFields() {
+        return new HashMap<>(documentFields);
+    }
+
```
Comment (Sandmannn) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: I am not very happy to introduce these getters, but they are required in `UpdateResponse`.  As of now, we have 3 different getters - 2 new for meta/nonmeta and 1 old for all aggregate fields

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 8]
```diff
@@ -222,20 +226,34 @@ public String sourceAsString() {
...
     public Map<String, DocumentField> getFields() {
+        Map<String, DocumentField> fields = new HashMap<>();
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: Do we want to create a new map every time we need to access `getFields` or `field`? May be better to create this `fields` map only once in the constructor?

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 9]
```diff
@@ -222,20 +226,34 @@ public String sourceAsString() {
         return sourceAsMap();
     }
 
+
+    public Map<String, DocumentField> getMetadataFields() {
+        return new HashMap<>(metaFields);
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: Do we need to create a new hashmap here or can we just return `metaFields`?

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 10]
```diff
@@ -222,20 +226,34 @@ public String sourceAsString() {
...
+    }
+
+    public Map<String, DocumentField> getDocumentFields() {
+        return new HashMap<>(documentFields);
+    }
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: Do we need to create a new hashmap here or can we just return `documentFields`?

Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: done

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 8 above)
Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: Actually I was asking something similar few patches ago, please see https://github.com/elastic/elasticsearch/pull/38373#discussion_r276022220
got the feeling that it is not the top concern for now, but I don't really have a strong opinion here.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 9 above)
Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: done

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 8 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: @Sandmannn  thanks, I agree with can optimize this in the subsequent PRs.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 11]
```diff
@@ -379,6 +385,37 @@ public static GetResult readGetResult(StreamInput in) throws IOException {
...
+            return;
+        }
+        for (Map.Entry<String, DocumentField> fieldEntry: fields.entrySet()) {
+            if (fieldEntry.getValue().isMetadataField()) {
+                outMetadata.put(fieldEntry.getKey(), fieldEntry.getValue());
+            } else {
+                outOther.put(fieldEntry.getKey(), fieldEntry.getValue());                
+            }
+        }
+
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: should we remove extra empty lines?

File: server/src/main/java/org/elasticsearch/index/get/ShardGetService.java [hunk 12]
```diff
@@ -201,9 +202,15 @@ private GetResult innerGetLoadFromStoredFields(String type, String id, String[]
 
             if (!fieldVisitor.fields().isEmpty()) {
                 fieldVisitor.postProcess(mapperService);
-                fields = new HashMap<>(fieldVisitor.fields().size());
+                nonMetaDataFields = new HashMap<>();
+                metaDataFields = new HashMap<>();
                 for (Map.Entry<String, List<Object>> entry : fieldVisitor.fields().entrySet()) {
-                    fields.put(entry.getKey(), new DocumentField(entry.getKey(), entry.getValue()));
+                    boolean isMetadataField = MapperService.isMetadataField(entry.getKey());
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: looks like `isMetadataField` variable is never used

File: server/src/test/java/org/elasticsearch/index/get/GetResultTests.java [hunk 13]
```diff
@@ -76,14 +76,15 @@ public void testToXContent() throws IOException {
         {
             GetResult getResult = new GetResult("index", "type", "id", 0, 1, 1, true, new BytesArray("{ \"field1\" : " +
                     "\"value1\", \"field2\":\"value2\"}"), singletonMap("field1", new DocumentField("field1",
-                    singletonList("value1"))));
+                    singletonList("value1"))),singletonMap("field1", new DocumentField("metafield",
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/index/get/GetResultTests.java]: nit. space before `singletonMap`

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 14]
```diff
@@ -379,6 +385,37 @@ public static GetResult readGetResult(StreamInput in) throws IOException {
...
+            fields = new HashMap<>(size);
+            for (int i = 0; i < size; i++) {
+                DocumentField field = DocumentField.readDocumentField(in);
+                fields.put(field.getName(), field);
+            }
+        }
+        return fields;
+    }
+    
+    public static void splitFieldsByMetadata(Map<String, DocumentField> fields, Map<String, DocumentField> outOther,
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: should be drop `public` modifier here?

**Review by mayya-sharipova — APPROVED** [review]
Thanks @Sandmannn. 
I have left some small comments, and provided they are addressed, and the build is green, this PR looks good to me.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 14 above)
Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: Unfortunately this method is required in tests e.g. https://github.com/elastic/elasticsearch/pull/38373/files#diff-219c73a7a0b84de3e96bdda87ec92270R218 when we are building random getResult using the builders of random DocumentFields and need to understand where to put each of the randomly generated fields. Not sure how to go around it.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 11 above)
Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: done

File: server/src/main/java/org/elasticsearch/index/get/ShardGetService.java (hunk 12 above)
Comment (Sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: done

File: server/src/test/java/org/elasticsearch/index/get/GetResultTests.java (hunk 13 above)
Comment (Sandmannn) [reply server/src/test/java/org/elasticsearch/index/get/GetResultTests.java]: done

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 14 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: @sandmannn removing `public` modifier will make this method package-private which will still allow it to be used in `GetResultTests` class.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 15]
```diff
@@ -222,20 +226,34 @@ public String sourceAsString() {
...
-        return fields.get(name);
+        return getFields().get(name);
     }
 
     @Override
     public Iterator<DocumentField> iterator() {
-        if (fields == null) {
+        // need to join the fields and metadata fields
+        Map<String, DocumentField> allFields = this.getFields();
+        if (allFields == null) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: I dont' think this is possible? `getFields()` never returns null

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 16]
```diff
@@ -244,21 +262,7 @@ public XContentBuilder toXContentEmbedded(XContentBuilder builder, Params params
...
-                }
-                if (field.isMetadataField()) {
-                    metaFields.add(field);
-                } else {
-                    otherFields.add(field);
-                }
-            }
-        }
-        for (DocumentField field : metaFields) {
+        for (DocumentField field : metaFields.values()) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: Using .values() can result in changes in ordering. However, it looks like we already had this problem. I opened https://github.com/elastic/elasticsearch/issues/42307 to followup.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 17]
```diff
@@ -393,15 +428,14 @@ public void readFrom(StreamInput in) throws IOException {
             if (source.length() == 0) {
                 source = null;
             }
-            int size = in.readVInt();
-            if (size == 0) {
-                fields = emptyMap();
+            if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/index/get/GetResult.java]: I think this should be 7.3.0, since we intend to backport to 7.x?

File: server/src/main/java/org/elasticsearch/index/get/ShardGetService.java [hunk 18]
```diff
@@ -201,9 +202,14 @@ private GetResult innerGetLoadFromStoredFields(String type, String id, String[]
 
             if (!fieldVisitor.fields().isEmpty()) {
                 fieldVisitor.postProcess(mapperService);
-                fields = new HashMap<>(fieldVisitor.fields().size());
+                nonMetaDataFields = new HashMap<>();
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: we should use `documentFields` for consistency

**Review by rjernst — APPROVED** [review]
Thank you for persevering on this issue @sandmannn! I just have a few more minor comments, which should be easy changes.

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 14 above)
Comment (sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: thanks, done

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 17 above)
Comment (sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: at the moment of writing this, there is only 7.2.0 https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/Version.java#L110 
Should we add 7.3.0 to that file?

File: server/src/main/java/org/elasticsearch/index/get/ShardGetService.java (hunk 18 above)
Comment (sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: done

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 15 above)
Comment (sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: done

File: server/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 17 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: Sorry I mispoke. 7.2.0 is good.

Comment (sandmannn) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: hm, what is the regular sequence of actions to create a downport here?

As expected, without adjustments in serialization for 7.2 the tests for the mixed cluster fail. Should I open a similar downport PR against https://github.com/elastic/elasticsearch/tree/7.x, get it merged, and then expect the build here to be green?

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: We handle this by temporarily disabling bwc tests. I just pushed the disabling to your branch. Tomorrow I will test the bwc behavior locally before pushing.

Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: @rjernst Looks like 7.2 has already feature freeze. Should this PR target 7.3 and master?

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/index/get/GetResult.java]: Yes, 7.2 was branched, but the 7.3 constant hasn't been pushed yet. I will update this branch today once the constant exists.

**Comment by rjernst** [comment]
I've tested backcompat after locally backporting and it looks good. I'll merge this once CI completes.

**Comment by rjernst** [comment]
Thanks for pushing through on this @sandmannn!!

**Comment by sandmannn** [comment]
thanks for merging! Now what about downport? Do we need a new PR against 7.x?

**Comment by rjernst** [comment]
@sandmannn I have already backported, see a49bafc.
